casanovo/denovo/dataloaders.py

```diff
@@ -64,24 +81,123 @@ def __init__(
+                        pa.string()),
+            CustomField("title",
+                        lambda x: x["params"]["title"],
+                        pa.string())
+        ]
+        self.custom_field_test_mzml = [
+            CustomField("scans", lambda x: x["id"], pa.string()),
+            CustomField("title", lambda x: x["id"], pa.string()),
+        ]
+        
+        self.custom_field_anno = [CustomField("seq", lambda x: x["params"]["seq"], pa.string())]
+
+    def make_dataset(self, paths, annotated, mode, shuffle):
+        """
+        Make spectrum datasets
+        Parameters
+        ----------
+        paths : Iterable[str]
+            Paths to input datasets
+        annotated: bool
+            True if peptide sequence annotations are available for the test
+            data.
+        mode: str {"train", "valid", "test"}
+            The mode indicating name of lance instance  
+        shuffle: bool
+            Indicates whether to shuffle training data based on buffer_size
+        """
+        custom_fields = self.custom_field_anno if annotated else []
+        
+        if mode=="test":
+            if all([Path(f).suffix in ('.mgf') for f in paths]):
+                custom_fields = custom_fields + self.custom_field_test_mgf
+            if all([Path(f).suffix in (".mzml",  ".mzxml", '.mzML') for f in paths]):
+                custom_fields = custom_fields + self.custom_field_test_mzml
+            
+        lance_path = f'{self.lance_dir}/{mode}.lance'
+        
+        parse_kwargs = dict(
+            preprocessing_fn=self.preprocessing_fn,
+            custom_fields=custom_fields,
+            valid_charge=self.valid_charge,
+
+        )
+
+        dataset_params = dict(
+            batch_size=self.train_batch_size if mode=="train" else self.eval_batch_size
+        )
+        anno_dataset_params = dataset_params | dict(
+            tokenizer=self.tokenizer,
+            annotations='seq',
+        )
+
+        if any([Path(f).suffix in (".lance") for f in paths]):
+            if annotated:
+                dataset = AnnotatedSpectrumDataset.from_lance(paths[0], **anno_dataset_params)
+            else:
+                dataset = SpectrumDataset.from_lance(paths[0], **dataset_params)
+        else:
+            if annotated:
+                dataset = AnnotatedSpectrumDataset(
+                    spectra=paths,
+                    path=lance_path,
+                    parse_kwargs=parse_kwargs,
+                    **anno_dataset_params,
+                )
+            else:
+                dataset = SpectrumDataset(
+                    spectra=paths,
+                    path=lance_path,
+                    parse_kwargs=parse_kwargs,
+                    **dataset_params,
+                )
+    
+        if shuffle:
+            dataset = ShufflerIterDataPipe(
+                dataset,
+                buffer_size=self.buffer_size
+            )
```

Comment:
I also found it very cool! With the current version of the training dataset in the mgf file the spectra are somehow sorted by peptide. That's why one for sure needs to set it up, as the training does not converge otherwise.